### ISSUE TITLE
Ignore "Expensive method invocation" if the code is wrapped inside `UNITY_EDITOR` preprocessor directive

### ISSUE THREAD

**Issue by AtaTrkgl** [JetBrains/resharper-unity#2345]
If an expensive method is wrapped inside `UNITY_EDITOR`, `UNITY_EDITOR_WIN`, `UNITY_EDITOR_OSX` or `UNITY_EDITOR_LINUX`, an _Expansive Method Invocation_ warning shouldn't be shown or at least state that it is for editor only.

Here is a reference to the Unity specific preprocessor statements: [Link](https://docs.unity3d.com/Manual/PlatformDependentCompilation.html)

Example:

The following method should not be marked as _Expansive Method Invocation_, as it won't make a difference when the project is built.
```csharp
private static void Method()
{
#if UNITY_EDITOR
        Debug.Log("Test Log.");
#endif
        // Do some stuff here (Non-Expansive)
}
```

However, the following code should be marked as _Expansive Method Invocation_, as `GetComponent` method is expansive.
```csharp
private static void Method()
{
#if UNITY_EDITOR
        Debug.Log("Test Log.");
#endif
        var x = GetComponent<SpriteRenderer>();
}
```

**Comment by Havie** [comment]
yes please.
when developing in editor a savvy dev will put editor only debug code inside of the if UNITY_EDITOR tags, this will never go into a release build. I do not need to be warned about expensive operations here, as they are already debug features not going into release.

**Comment by Havie** [comment]
Another option is to allow devs to turn off warnings for "Debug.Logs" as a whole if thats a simpler fix

**Comment by synthc** [comment]
It doesn't really make sense for this warning to trigger on anything from the Debug class.

**Comment by AtaTrkgl** [comment]
> It doesn't really make sense for this warning to trigger on anything from the Debug class; as you're presumed to have removed all of that on release anyway.

Nope, when you have a release-ready build, what makes sense is to wrap the necessary debug code with `UNITY_EDITOR` preprocessor directives so you can still benefit from the debug system in the editor while your builds are safe to ship from performance issues.

**Comment by synthc** [comment]
> Nope, when you have a release-ready build, what makes sense is to wrap the necessary debug code with `UNITY_EDITOR` preprocessor directives so you can still benefit from the debug system in the editor while your builds are safe to ship from performance issues.

Sure, any logging/debugging systems you want to leave in would be wrapped in an #if Unity_Editor directive.  Any temporary debug code should just be removed.

In either case, there won't be anything from the Debug class in the release build, so I don't want to see performance warnings for that class.  I don't want to take the time to put directives on every temporary Debug.Log and having every function with a Debug.Log call be marked as expensive makes this feature too noisy to be usable.

I totally agree that this shouldn't trigger on anything with `UNITY_EDITOR` directive either.

**Comment by guplem** [comment]
> allow devs to turn off warnings for "Debug.Logs" as a whole if thats a simpler fix

I am fully onboard with this. As a matter of fact, it might be even intentional for the devs to leave the `Debug.Log`, since those can be captured and used while testing. Performance will be penalized, but at least you have logs...

Also, I assume here everyone is talking about `Debug.Log`, `Debug.LogError`, `Debug.LogWarning`, ...

**Comment by morfeyc** [comment]
> allow devs to turn off warnings for "Debug.Logs" as a whole if thats a simpler fix

+1 on this

**Comment by settenop** [comment]
I know this is not a definitive solution for the UNITY_EDITOR directive, but I'm using this to avoid the expensive method invocation warnings related to Debug.Log

```csharp
using System.Diagnostics;
using System.Reflection;
using Debug = UnityEngine.Debug;
using Object = UnityEngine.Object;

namespace DebugHelpers
{
    public static class Devug
    {
        private static readonly MethodInfo LogMethod;
        private static readonly MethodInfo WarnMethod;
        private static readonly MethodInfo ErrMethod;

        static Devug()
        {
            var type = typeof(Debug);
            LogMethod = type.GetMethod("Log", new[] { typeof(object), typeof(Object) });
            WarnMethod = type.GetMethod("LogWarning", new[] { typeof(object), typeof(Object) });
            ErrMethod = type.GetMethod("LogError", new[] { typeof(object), typeof(Object) });
        }
        
        [Conditional("UNITY_EDITOR"), Conditional("DEVELOPMENT_BUILD")]
        public static void Log(object message, Object context = null)
        {
            LogMethod.Invoke(null, new [] { message, context });
        }

        [Conditional("UNITY_EDITOR"), Conditional("DEVELOPMENT_BUILD")]
        public static void LogWarning(object message, Object context = null)
        {
            WarnMethod.Invoke(null, new[] { message, context });
        }

        [Conditional("UNITY_EDITOR"), Conditional("DEVELOPMENT_BUILD")]
        public static void LogError(object message, Object context = null)
        {
            ErrMethod.Invoke(null, new[] { message, context });
        }

        [Conditional("UNITY_EDITOR"), Conditional("DEVELOPMENT_BUILD")]
        public static void LogSuccess(object message, Object context = null)
        {
            Log($"<b><color=#4CAF50>[SUCCESS]</color></b> {message}", context);
        }

        [Conditional("UNITY_EDITOR"), Conditional("DEVELOPMENT_BUILD")]
        public static void LogInfo(object message, Object context = null)
        {
            Log($"<b><color=#2196F3>[INFO]</color></b> {message}", context);
        }

        [Conditional("UNITY_EDITOR"), Conditional("DEVELOPMENT_BUILD")]
        public static void LogPhys(object message, Object context = null)
        {
            Log($"<b><color=#FF9800>[PHYSICS]</color></b> {message}", context);
        }
    }
}
```
